README.md
# It converts date from english to nepali and vice-versa

[![Latest Version on Packagist](https://img.shields.io/packagist/v/nisshan/nepali-calendar.svg?style=flat-square)](https://packagist.org/packages/nisshan/nepali-calendar)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/nisshan/nepali-calendar/run-tests?label=tests)](https://github.com/nisshan/nepali-calendar/actions?query=workflow%3Arun-tests+branch%3Amain)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/nisshan/nepali-calendar/Check%20&%20fix%20styling?label=code%20style)](https://github.com/nisshan/nepali-calendar/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amain)
[![Total Downloads](https://img.shields.io/packagist/dt/nisshan/nepali-calendar.svg?style=flat-square)](https://packagist.org/packages/nisshan/nepali-calendar)

This is a laravel package that converts English date to Nepali and vice versa.

## Installation

You can install the package via composer:

```bash
composer require nisshan/nepali-calendar
```

You can publish the config file with:

```bash
php artisan vendor:publish --tag="nepali-calendar-config"
```

This is the contents of the published config file:

```php
return [
    'date-format' => 'Y m, d is default format',
    
    'nepali-dates' => 'array of nepali dates'
];
```

## Usage

```php
toNepaliDate('2022-1-17'); => '2078 10, 03'

toNepaliDate('2022-1-17', 'Y, m d, D'); => '2078, 10 03, Monday'
```

to convert to Nepali Date from English and

```php
toEnglishDate('2078-10-03'); => '2022 1, 17'

toEnglishDate('2078-10-03', 'Y, m d, D'); => '2022, 1 17, Monday'
```

to convert to English from Nepali date. 
The date Format parameter is optional which format the date if not passed will be used from configuration file. 
The above two implements shows how you can pass the english date to the function with or without a separator that
changes helps in separating the date using Helper file or you can also convert the date as

```php

use Nisshan\NepaliCalendar\DateConversion;

$nepaliDate = DateConversion::convert('2022-1-17')->toNepali();
$nepaliDateWithFormat = DateConversion::convert('2022-1-17','Y, m d, D')->toNepali();


$englishDate = DateConversion::convert('2078-10-03')->toNepali();
$englishWithFormat = DateConversion::convert('2078-10-03','Y, m d, D')->toNepali();

```
It will give same output as helper function.

## Format that you can use

currently, I have used ['Y','M','D','d','m','y'] as date formatter you can use any combination of the following to output formatted date.


```php

  'Y','Y'  => 'Year',
  'M'      => 'Month Name',
  'D'      => 'Name of Week Day',
  'd'      => 'date',
  'm'      => 'month'

```

## Roadmap for next version

I have worked in 2-3 system where nepali date is required, and I only had to convert to and from dates so currently only had that function but I have some other use case that might be helpful to another projects and myself in near future so will think on them and work on them

- Find better way to format the date
- Converting dates to nepali(font)

## Thank you note

If you are here by any chance and using my package I would like to thank you and would love to accept  PR on improvement of code or new features. Thank you.

## Testing

```bash
composer test
```

## Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information on what has changed recently.

## Contributing

Please see [CONTRIBUTING](.github/CONTRIBUTING.md) for details.

## Security Vulnerabilities

Please review [our security policy](../../security/policy) on how to report security vulnerabilities.

## Credits

- [nisshan](https://github.com/Nisshan)
- [All Contributors](../../contributors)

## License

The MIT License (MIT). Please see [License File](LICENSE.md) for more information.


CHANGELOG.md
# Changelog

All notable changes to `nepali-calendar` will be documented in this file.

## 1.0.0 - 2022-01-23

initial release

## 1.0.0 - 2022-01-23

- initial release


tests/Pest.php
<?php

use Nisshan\NepaliCalendar\Tests\TestCase;



uses(TestCase::class)->in(__DIR__);






tests/TestCase.php
<?php

namespace Nisshan\NepaliCalendar\Tests;

use Illuminate\Database\Eloquent\Factories\Factory;
use Nisshan\NepaliCalendar\NepaliCalendarServiceProvider;
use Orchestra\Testbench\TestCase as Orchestra;

class TestCase extends Orchestra
{
    protected array $dates ;

    protected function setUp(): void
    {
        parent::setUp();
        $this->dates = config('nepali-calendar.nepali-dates');
        Factory::guessFactoryNamesUsing(
            fn (string $modelName) => 'Nisshan\\NepaliCalendar\\Database\\Factories\\'.class_basename($modelName).'Factory'
        );
    }

    protected function getPackageProviders($app)
    {
        return [
            NepaliCalendarServiceProvider::class,
        ];
    }

    public function getEnvironmentSetUp($app)
    {
        config()->set('database.default', 'testing');

        /*
        $migration = include __DIR__.'/../database/migrations/create_nepali-calendar_table.php.stub';
        $migration->up();
        */
    }

    public function getFirstNepaliInValidYear(): int
    {
        return $this->dates[0][0] - 1;
    }

    public function getLastNepaliInValidYear(): int
    {
        return end($this->dates)[0] + 1;
    }
}


tests/DateConversionTest.php
<?php

use Nisshan\NepaliCalendar\DateConversion;

it('converts to default format when format is not passed', function () {
    $nepaliDate = DateConversion::convert('2022-1-17')->toNepali();
    $this->assertEquals('2078 10, 03', $nepaliDate);
});

it('converts to date in same format as passed in', function () {
    $nepaliDate = DateConversion::convert('2022-1-17', 'Y M d')->toNepali();
    $this->assertEquals('2078 Magh 03', $nepaliDate);
});

it('converts date using helper function with default separator', function () {
    $nepaliDate = toNepaliDate('2022-1-17');
    $this->assertEquals('2078 10, 03', $nepaliDate);
});

it('converts to date in y/m/d format with custom separator using helper function', function () {
    $nepaliDate = toNepaliDate('2022-1-17', 'Y, m d, D');
    $this->assertEquals('2078, 10 03, Monday', $nepaliDate);
});


src/DateConversion.php
<?php

namespace Nisshan\NepaliCalendar;

class DateConversion
{
    private string $date;
    private string|null $format;

    public function __construct(string $date, string|null $format)
    {
        $this->date = $date;
        $this->format = $format;
    }

    public static function convert($date, $format = null): self
    {
        return new static($date, $format);
    }

    public function toNepali(): string
    {
        $date = $this->changeDateFormat();
        $calendar = new NepaliCalendar();
        $nepaliDate = $calendar->englishToNepali($date[0], $date[1], $date[2]);

        return $this->formatDate($nepaliDate);
    }

    public function toEnglish(): string
    {
        $date = $this->changeDateFormat();
        $calendar = new NepaliCalendar();
        $englishDate = $calendar->nepaliToEnglish($date[0], $date[1], $date[2]);

        return $this->formatDate($englishDate);
    }

    private function formatDate($date): string
    {
        $dateFormat = $this->explodeFormat();

        return collect($dateFormat)->reduce(function ($formatedDate, $part) use ($date) {
            return $formatedDate . $this->dateLookupTable($date, $part);
        });
    }

    private function changeDateFormat(): array
    {
        return explode('-', $this->date);
    }

    private function explodeFormat(): array
    {
        $format = $this->format ?: config('nepali-calendar.date-format');

        return str_split($format);
    }

    private function dateLookupTable($date, $format): string|null
    {
        if (! in_array($format, ['Y','M','D','d','m','y'])) {
            return $format;
        }

        return [
            'Y' => $date['year'],
            'y' => $date['year'],
            'M' => $date['mname'],
            'D' => $date['day'],
            'd' => $date['date'],
            'm' => $date['month'],
        ][$format];
    }
}


config/nepali-calendar.php
<?php
// config for Nisshan/NepaliCalendar
return [

    'date-format' => 'Y m, d',

    'nepali-dates' => [
        0 => [2000, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        1 => [2001, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        2 => [2002, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        3 => [2003, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        4 => [2004, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        5 => [2005, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        6 => [2006, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        7 => [2007, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        8 => [2008, 31, 31, 31, 32, 31, 31, 29, 30, 30, 29, 29, 31],
        9 => [2009, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        10 => [2010, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        11 => [2011, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        12 => [2012, 31, 31, 31, 32, 31, 31, 29, 30, 30, 29, 30, 30],
        13 => [2013, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        14 => [2014, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        15 => [2015, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        16 => [2016, 31, 31, 31, 32, 31, 31, 29, 30, 30, 29, 30, 30],
        17 => [2017, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        18 => [2018, 31, 32, 31, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        19 => [2019, 31, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        20 => [2020, 31, 31, 31, 32, 31, 31, 30, 29, 30, 29, 30, 30],
        21 => [2021, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        22 => [2022, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 30],
        23 => [2023, 31, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        24 => [2024, 31, 31, 31, 32, 31, 31, 30, 29, 30, 29, 30, 30],
        25 => [2025, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        26 => [2026, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        27 => [2027, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        28 => [2028, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        29 => [2029, 31, 31, 32, 31, 32, 30, 30, 29, 30, 29, 30, 30],
        30 => [2030, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        31 => [2031, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        32 => [2032, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        33 => [2033, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        34 => [2034, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        35 => [2035, 30, 32, 31, 32, 31, 31, 29, 30, 30, 29, 29, 31],
        36 => [2036, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        37 => [2037, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        38 => [2038, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        39 => [2039, 31, 31, 31, 32, 31, 31, 29, 30, 30, 29, 30, 30],
        40 => [2040, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        41 => [2041, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        42 => [2042, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        43 => [2043, 31, 31, 31, 32, 31, 31, 29, 30, 30, 29, 30, 30],
        44 => [2044, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        45 => [2045, 31, 32, 31, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        46 => [2046, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        47 => [2047, 31, 31, 31, 32, 31, 31, 30, 29, 30, 29, 30, 30],
        48 => [2048, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        49 => [2049, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 30],
        50 => [2050, 31, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        51 => [2051, 31, 31, 31, 32, 31, 31, 30, 29, 30, 29, 30, 30],
        52 => [2052, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        53 => [2053, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 30],
        54 => [2054, 31, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        55 => [2055, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        56 => [2056, 31, 31, 32, 31, 32, 30, 30, 29, 30, 29, 30, 30],
        57 => [2057, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        58 => [2058, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        59 => [2059, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        60 => [2060, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        61 => [2061, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        62 => [2062, 30, 32, 31, 32, 31, 31, 29, 30, 29, 30, 29, 31],
        63 => [2063, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        64 => [2064, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        65 => [2065, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        66 => [2066, 31, 31, 31, 32, 31, 31, 29, 30, 30, 29, 29, 31],
        67 => [2067, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        68 => [2068, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        69 => [2069, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        70 => [2070, 31, 31, 31, 32, 31, 31, 29, 30, 30, 29, 30, 30],
        71 => [2071, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        72 => [2072, 31, 32, 31, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        73 => [2073, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 31],
        74 => [2074, 31, 31, 31, 32, 31, 31, 30, 29, 30, 29, 30, 30],
        75 => [2075, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        76 => [2076, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 30],
        77 => [2077, 31, 32, 31, 32, 31, 30, 30, 30, 29, 30, 29, 31],
        78 => [2078, 31, 31, 31, 32, 31, 31, 30, 29, 30, 29, 30, 30],
        79 => [2079, 31, 31, 32, 31, 31, 31, 30, 29, 30, 29, 30, 30],
        80 => [2080, 31, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 30],
        81 => [2081, 31, 31, 32, 32, 31, 30, 30, 30, 29, 30, 30, 30],
        82 => [2082, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 30, 30],
        83 => [2083, 31, 31, 32, 31, 31, 30, 30, 30, 29, 30, 30, 30],
        84 => [2084, 31, 31, 32, 31, 31, 30, 30, 30, 29, 30, 30, 30],
        85 => [2085, 31, 32, 31, 32, 30, 31, 30, 30, 29, 30, 30, 30],
        86 => [2086, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 30, 30],
        87 => [2087, 31, 31, 32, 31, 31, 31, 30, 30, 29, 30, 30, 30],
        88 => [2088, 30, 31, 32, 32, 30, 31, 30, 30, 29, 30, 30, 30],
        89 => [2089, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 30, 30],
        90 => [2090, 30, 32, 31, 32, 31, 30, 30, 30, 29, 30, 30, 30],
        91 => [2091, 31, 31, 32, 32, 31, 30, 30, 29, 30, 29, 30, 30],
        92 => [2092, 30, 31, 32, 32, 31, 30, 30, 30, 29, 30, 30, 30],
        93 => [2093, 30, 32, 31, 32, 31, 30, 30, 30, 29, 29, 30, 30],
        94 => [2094, 31, 31, 32, 31, 31, 30, 30, 30, 29, 30, 30, 30],
        95 => [2095, 31, 31, 32, 31, 31, 31, 30, 29, 30, 30, 30, 30]
    ]

];


src/NepaliCalendar.php
<?php

namespace Nisshan\NepaliCalendar;

use InvalidArgumentException;

class NepaliCalendar
{
    private array $nepaliDates;

    public function __construct()
    {
        $this->nepaliDates = config('nepali-calendar.nepali-dates');
    }

    private array $nepaliDate = ['year' => '', 'month' => '', 'date' => '', 'day' => '', 'mname' => '', 'num_day' => ''];
    private array $englishDate = ['year' => '', 'month' => '', 'date' => '', 'day' => '', 'mname' => '', 'num_day' => ''];

    public function englishToNepali(int $yy, int $mm, int $dd): array
    {
        // Check for date range
        $checkRange = $this->isInEnglishRange($yy, $mm, $dd);

        if ($checkRange !== true) {
            die($checkRange);
        }

        // Month data.
        $month = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
        // Month for leap year
        $lmonth = [31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];

        $def_eyy = 1944;    // initial english year.
        $def_nyy = 2000;    // initial nepali date in dataset
        $def_nmm = 9;       // poush => Jan

        $def_ndd = 17 - 1;  // initial nepali date.
        $total_eDays = 0;
        $day = 7 - 1;

        // Count total no. of days in-terms of  year
        for ($i = 0; $i < ($yy - $def_eyy); $i++) { //total days for month calculation...(english)
            if ($this->isLeapYear($def_eyy + $i) === true) {
                for ($j = 0; $j < 12; $j++) {
                    $total_eDays += $lmonth[$j];
                }
            } else {
                for ($j = 0; $j < 12; $j++) {
                    $total_eDays += $month[$j];
                }
            }
        }


        // Count total no. of days in-terms of month
        for ($i = 0; $i < ($mm - 1); $i++) {
            if ($this->isLeapYear($yy) === true) {
                $total_eDays += $lmonth[$i];
            } else {
                $total_eDays += $month[$i];
            }
        }

        // Count total no. of days in-terms of date
        $total_eDays += $dd;

        $i = 0;         // year array key
        $j = $def_nmm; // month array key for year
        $total_nDays = $def_ndd;

        $m = $def_nmm;
        $y = $def_nyy;

        // Count nepali date from array
        while ($total_eDays != 0) {
            $monthDaysOfYear = $this->nepaliDates[$i][$j];

            $total_nDays++;     //count the days
            $day++;             //count the days in terms of 7 days
            if ($total_nDays > $monthDaysOfYear) {
                $m++;
                $total_nDays = 1;
                $j++;
            }

            if ($day > 7) {
                $day = 1;
            }

            if ($m > 12) {
                $y++;
                $m = 1;
            }

            if ($j > 12) {
                $j = 1;
                $i++;
            }

            $total_eDays--;
        }

        $numDay = $day;
        $this->nepaliDate['year'] = $y;
        $this->nepaliDate['month'] = $m <= 9 ? "0" . $m : $m;
        $this->nepaliDate['date'] = $total_nDays <= 9 ? "0" . $total_nDays : $total_nDays;
        $this->nepaliDate['day'] = $this->getDateOfWeek($day);
        $this->nepaliDate['mname'] = $this->getNepaliMonthName($m);
        $this->nepaliDate['num_day'] = $numDay;

        return $this->nepaliDate;
    }

    public function nepaliToEnglish(int $yy, int $mm, int $dd): array
    {
        $def_eyy = 1943;
        $def_emm = 4;
        $def_edd = 14 - 1;  // initial english date.
        $def_nyy = 2000;
        $total_nDays = 0;
        $day = 4 - 1;
        $k = 0;
        $month = [0, 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
        $lmonth = [0, 31, 29, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
        // Check for date range
        $check = $this->isInNepaliRange($yy, $mm, $dd);
        if ($check !== true) {
            die($check);
        }

        // Count total days in-terms of year
        for ($i = 0; $i < ($yy - $def_nyy); $i++) {
            for ($j = 1; $j <= 12; $j++) {
                $total_nDays += $this->nepaliDates[$k][$j];
            }
            $k++;
        }
        // Count total days in-terms of month
        for ($j = 1; $j < $mm; $j++) {
            $total_nDays += $this->nepaliDates[$k][$j];
        }
        // Count total days in-terms of dat
        $total_nDays += $dd;
        // Calculation of equivalent english date...
        $total_eDays = $def_edd;
        $m = $def_emm;
        $y = $def_eyy;
        while ($total_nDays != 0) {
            if ($this->isLeapYear($y)) {
                $a = $lmonth[$m];
            } else {
                $a = $month[$m];
            }
            $total_eDays++;
            $day++;
            if ($total_eDays > $a) {
                $m++;
                $total_eDays = 1;
                if ($m > 12) {
                    $y++;
                    $m = 1;
                }
            }
            if ($day > 7) {
                $day = 1;
            }
            $total_nDays--;
        }

        $numDay = $day;
        $this->englishDate['year'] = $y;
        $this->englishDate['month'] = $m;
        $this->englishDate['date'] = $total_eDays;
        $this->englishDate['day'] = $this->getDateOfWeek($day);
        $this->englishDate['mname'] = $this->getEnglishMonth($m);
        $this->englishDate['num_day'] = $numDay;

        return $this->englishDate;
    }

    private function getDateOfWeek(int $day): string
    {
        return[
            1 => "Sunday",
            2 => "Monday",
            3 => "Tuesday",
            4 => "Wednesday",
            5 => "Thursday",
            6 => "Friday",
            7 => "Saturday",
        ][$day];
    }

    private function getEnglishMonth(int $month): string
    {
        return [
            1 => "January",
            2 => "February",
            3 => "March",
            4 => "April",
            5 => "May",
            6 => "June",
            7 => "July",
            8 => "August",
            9 => "September",
            10 => "October",
            11 => "November",
            12 => "December",
        ][$month];
    }

    private function getNepaliMonthName(int $month): string
    {
        return [
            1 => "Baishak",
            2 => "Jestha",
            3 => "Ashad",
            4 => "Shrawn",
            5 => "Bhadra",
            6 => "Ashwin",
            7 => "kartik",
            8 => "Mangshir",
            9 => "Poush",
            10 => "Magh",
            11 => "Falgun",
            12 => "Chaitra",
        ][$month];
    }

    private function isInEnglishRange(int $yy, int $mm, int $dd): bool
    {
        if ($yy < $this->getFirstEnglishInvalidYear() || $yy > $this->getLastEnglishInvalidYear()) {
            throw new InvalidArgumentException('Invalid Date Provided');
        }
        if ($mm < 1 || $mm > 12) {
            throw new InvalidArgumentException('Month can only contain value from 1-12');
        }
        if ($dd < 1 || $dd > 31) {
            throw new InvalidArgumentException('English Day can only contain value from 1-31');
        }

        return true;
    }

    private function isInNepaliRange(int $yy, int $mm, int $dd): bool
    {
        if ($yy < $this->getFirstNepaliInValidYear() || $yy > $this->getLastNepaliInValidYear()) {
            throw new InvalidArgumentException('Invalid Date Provided');
        }
        if ($mm < 1 || $mm > 12) {
            throw new InvalidArgumentException('Month can only contain value from 1-12');
        }
        if ($dd < 1 || $dd > 32) {
            throw new InvalidArgumentException('Nepali Day can only contain value from 1-32');
        }

        return true;
    }

    private function isLeapYear(int $year): bool
    {
        if ($year % 100 == 0) {
            if ($year % 400 == 0) {
                return true;
            } else {
                return false;
            }
        } else {
            if ($year % 4 == 0) {
                return true;
            } else {
                return false;
            }
        }
    }

    private function getLastNepaliInValidYear(): int
    {
        return end($this->nepaliDates)[0];
    }

    private function getFirstNepaliInValidYear(): int
    {
        return $this->nepaliDates[0][0];
    }

    private function getLastEnglishInvalidYear(): int
    {
        $date = $this->nepaliToEnglish($this->getLastNepaliInValidYear(), 1, 1);

        return $date['year'];
    }

    private function getFirstEnglishInvalidYear(): int
    {
        $date = $this->nepaliToEnglish($this->getFirstNepaliInValidYear(), 1, 1);

        return $date['year'];
    }
}
